src/cpu/x64/jit_uni_reorder.cpp

```diff
@@ -902,6 +904,341 @@ struct jit_uni_reorder_kernel_f32 : public kernel_t, public jit_generator {
+            int in_tail, int out_tail) {
+        gen_tr8x8(i_off, o_off, input_stride, output_stride, in_tail, out_tail);
+    }
+
+    void gen_ker16x16_in_8x8(int input_stride, int output_stride) {
+        const auto lane = 16;
+        const auto sub_lane = lane / 2;
+        gen_ker8x8(0, 0, input_stride, output_stride, sub_lane, sub_lane);
+        gen_ker8x8(input_stride * sub_lane * itype_sz, sub_lane * otype_sz,
+                input_stride, output_stride, sub_lane, sub_lane);
+        gen_ker8x8(sub_lane * itype_sz, output_stride * sub_lane * otype_sz,
+                input_stride, output_stride, sub_lane, sub_lane);
+        gen_ker8x8((input_stride * sub_lane + sub_lane) * itype_sz,
+                (output_stride * sub_lane + sub_lane) * otype_sz, input_stride,
+                output_stride, sub_lane, sub_lane);
+    }
+
+    // tail can be 1 ~ 16, using avx2 for now
+    void gen_ker16x16_in_8x8(
+            int input_stride, int output_stride, int in_tail, int out_tail) {
+        constexpr auto lane = 16;
+        constexpr auto sub_lane = lane / 2;
+        auto tail = in_tail != lane ? in_tail : out_tail;
+
+        const auto l_tail = tail < sub_lane ? tail : sub_lane;
+        const auto u_tail = tail < sub_lane ? 0 : tail - sub_lane;
+
+        if (tail == in_tail) {
+            gen_ker8x8(0, 0, input_stride, output_stride, l_tail, sub_lane);
+            gen_ker8x8(input_stride * sub_lane * itype_sz, sub_lane * otype_sz,
+                    input_stride, output_stride, l_tail, sub_lane);
+            gen_ker8x8(sub_lane * itype_sz, output_stride * sub_lane * otype_sz,
+                    input_stride, output_stride, u_tail, sub_lane);
+            gen_ker8x8(itype_sz * (input_stride * sub_lane + sub_lane),
+                    otype_sz * (output_stride * sub_lane + sub_lane),
+                    input_stride, output_stride, u_tail, sub_lane);
+        } else {
+            gen_ker8x8(0, 0, input_stride, output_stride, sub_lane, l_tail);
+            gen_ker8x8(input_stride * sub_lane * itype_sz, sub_lane * otype_sz,
+                    input_stride, output_stride, sub_lane, u_tail);
+            gen_ker8x8(sub_lane * itype_sz, output_stride * sub_lane * itype_sz,
+                    input_stride, output_stride, sub_lane, l_tail);
+            gen_ker8x8(itype_sz * (input_stride * sub_lane + sub_lane),
+                    otype_sz * (output_stride * sub_lane + sub_lane),
+                    input_stride, output_stride, sub_lane, u_tail);
+        }
+    }
+
+    inline void operator()(
+            const void *in, void *out, const float *scale, bool tail) const {
+        if (tail)
+            ker_tail_(in, out, scale);
+        else
+            ker_(in, out, scale);
+    }
+
+    inline void operator()(
+            const void *in, void *out, const float *scale) const {
+        ker_(in, out, scale);
+    }
+
+    // Mask is fixed inside code
+    inline void set_mask() { set_mask_(); }
```

Comment:
I don't think this is a reliable way to set a mask between jit-kernels. The compiler is not obligated to preserve the xmm registers between the function calls (even if the go one after another...)

Reply:
For Linux, a __asm__ volatile (:::"xmm15"); would have some effect on this?